cpp/include/cudf/ast/detail/transform.cuh

```diff
@@ -177,241 +422,437 @@ struct row_evaluator {
- * and copies that to the device with a single host-device memory copy. Because the plan tends to be
- * small, this is the most efficient approach for low latency.
- *
- */
-struct ast_plan {
-  ast_plan(linearizer const& expr_linearizer,
-           rmm::cuda_stream_view stream,
-           rmm::mr::device_memory_resource* mr)
-    : _sizes{}, _data_pointers{}
-  {
-    add_to_plan(expr_linearizer.data_references());
-    add_to_plan(expr_linearizer.literals());
-    add_to_plan(expr_linearizer.operators());
-    add_to_plan(expr_linearizer.operator_source_indices());
+ private:
+  /**
+   * @brief Helper struct for type dispatch on the result of an expression.
+   *
+   * Evaluating an expression requires multiple levels of type dispatch to
+   * determine the input types, the operation type, and the output type. This
+   * helper class is a functor that handles the operator dispatch, invokes the
+   * operator, and dispatches output writing based on the resulting data type.
+   */
+  struct expression_output_handler {
+   public:
+    __device__ expression_output_handler(expression_evaluator<has_nulls> const& evaluator)
+      : evaluator(evaluator)
+    {
+    }
 
-    // Create device buffer
-    auto const buffer_size = std::accumulate(_sizes.cbegin(), _sizes.cend(), 0);
-    auto buffer_offsets    = std::vector<int>(_sizes.size());
-    thrust::exclusive_scan(_sizes.cbegin(), _sizes.cend(), buffer_offsets.begin(), 0);
+    /**
+     * @brief Resolves an output data reference and assigns result value.
+     *
+     * Only output columns (COLUMN) and intermediates (INTERMEDIATE) are supported as output
+     * reference types. Intermediates must be of fixed width less than or equal to
+     * sizeof(std::int64_t). This requirement on intermediates is enforced by the linearizer.
+     *
+     * @tparam Element Type of result element.
+     * @tparam OutputType The container type that data will be inserted into.
+     *
+     * @param output_object The container that data will be inserted into.
+     * @param device_data_reference Data reference to resolve.
+     * @param row_index Row index of data column.
+     * @param result Value to assign to output.
+     */
+    template <typename Element,
+              typename OutputType,
+              CUDF_ENABLE_IF(is_rep_layout_compatible<Element>())>
+    __device__ void resolve_output(OutputType& output_object,
+                                   const detail::device_data_reference device_data_reference,
+                                   const cudf::size_type row_index,
+                                   const possibly_null_value_t<Element, has_nulls> result) const
+    {
+      auto const ref_type = device_data_reference.reference_type;
+      if (ref_type == detail::device_data_reference_type::COLUMN) {
+        output_object.template set_value<Element>(row_index, result);
+      } else {  // Assumes ref_type == detail::device_data_reference_type::INTERMEDIATE
```

Comment:
Couldn't the `reference_type` also be a `value_expression_result`?

Reply:
No, but I think what's probably confusing you is definitely a weakness with the current design. I suspect that you came to this conclusion because the conditional in the `if` checks whether the output reference is a `COLUMN`, correct? If not, then feel free to skip the rest of this message.

If so, then the problem is that the `COLUMN` reference type here is being used to indicate any non-intermediate output, i.e. `output_object` in line 682 may be any subclass of `expression_result`. `device_data_reference_type` encapsulates whether a piece of data is a column, a literal, or an intermediate value, but without any reference to whether it's an input or an output. The distinction between input and output is given by `device_data_reference::table_reference`, which is only meaningful when `device_data_reference::reference_type == COLUMN` and indicates either `LEFT`, `RIGHT`, or `OUTPUT`. These choices don't provide a really meaningful way to indicate what the output type is, so the code currently handles that by passing the appropriate type of `output_object` through so that on line 682 the output is placed in the right location.

This PR is designed to maximize the separation between the parsing of an expression and its evaluation on a set of operands with a particular output destination, but this issue is one example of where those lines get fuzzy. I considered adding a new `device_data_reference_type::OUTPUT` as a way to indicate the output data, but I don't particularly like that because whether data is input or output is semantically different from the information currently encoded by the reference type. `OUTPUT` and `COLUMN` wouldn't be mutually exclusively, and you would have a situation where input references are not explicitly indicated as such while outputs would be. On the other hand, it would more clearly isolate the expression parsing from the evaluation, and maybe it would be clearer to future developers. If you would prefer that change (or have an alternate proposal) I'm happy to change this.